Jetpack Connection: do it via iframe

roccotripaldi opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

Fixes no known issues

#### Changes proposed in this Pull Request:
If you have a constant set, you'll connect via iFrame within
wp-admin instead of being sent to WordPress.com

This is the last bit of #13112 

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Master Thread: p1HpG7-7nj-p2

#### Testing instructions:
* Run this branch on an unconnected site
* In your wp-config.php, or somewhere similar, add this:
`define ( 'JETPACK_SHOULD_USE_CONNECTION_IFRAME', true );`
* Try connecting and notice the iframe-y newness!
* Connections without the constant should continue to work as usual

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* None

roccotripaldi committed:
Jetpack Connection: do it via iframe

If you have a constant set, you'll connect via iFrame within
wp-admin instead of being sent to WordPress.com

lezama commented on _inc/connect-button.js:
```diff
@@ -1,6 +1,57 @@
+		handleClick: function() {
+			jetpackConnectButton.isRegistering = true;
+			$( '.jp-connect-button' ).text( jpConnect.buttonTextRegistering );
+			console.log( 'sending request to', jpConnect.apiBaseUrl + '/connection/register' );
```
```suggestion
```

lezama commented on _inc/connect-button.js:
```diff
@@ -1,6 +1,57 @@
+					_wpnonce: jpConnect.apiNonce,
+				},
+				error: function( error ) {
+					console.log( 'request failed' );
```
```suggestion
```

lezama commented on _inc/connect-button.js:
```diff
@@ -1,6 +1,57 @@
+				},
+				error: function( error ) {
+					console.log( 'request failed' );
+					console.log( error );
```
let's print a proper message with `console.warn`

lezama commented on _inc/connect-button.js:
```diff
@@ -1,6 +1,57 @@
+					jetpackConnectButton.isRegistering = false;
+				},
+				success: function( data ) {
+					console.log( 'request success' );
```
```suggestion
```

lezama commented on _inc/connect-button.js:
```diff
@@ -1,6 +1,57 @@
+		},
+		receiveData: function( event ) {
+			if ( event.origin === 'https://jetpack.wordpress.com' ) {
+				console.log( 'got message', event );
```
```suggestion
```

lezama commented on _inc/connect-button.js:
```diff
@@ -1,6 +1,57 @@
+			} );
+		},
+		receiveData: function( event ) {
+			if ( event.origin === 'https://jetpack.wordpress.com' ) {
```
should we use APIBASE or something alike?

roccotripaldi replied:
yes we should. that's likely why @tyxla was not able to redirect during his flow

tyxla replied:
Nice catch!

lezama commented on _inc/connect-button.js:
```diff
@@ -1,6 +1,57 @@
+		receiveData: function( event ) {
+			if ( event.origin === 'https://jetpack.wordpress.com' ) {
+				console.log( 'got message', event );
+				// todo: && e.source === this.iframe.contentWindow
```
let's add this back

lezama commented on _inc/connect-button.js:
```diff
@@ -1,6 +1,57 @@
+			}
+		},
+		handleAuthorizationComplete: function() {
+			console.log( 'finishing auth' );
```
```suggestion
```

lezama commented on class.jetpack-connection-banner.php:
```diff
@@ -134,13 +134,33 @@ public static function enqueue_connect_button_scripts() {
+			array(
+				'apiBaseUrl' => site_url('/wp-json/jetpack/v4'),
+				'registrationNonce' => wp_create_nonce( 'jetpack-registration-nonce' ),
+                'apiNonce' => wp_create_nonce( 'wp_rest' ),
```
indentation looks wonky

roccotripaldi committed:
Update _inc/connect-button.js

Co-Authored-By: Miguel Lezama <miguel@automattic.com>

roccotripaldi committed:
Update _inc/connect-button.js

Co-Authored-By: Miguel Lezama <miguel@automattic.com>

roccotripaldi committed:
Update _inc/connect-button.js

Co-Authored-By: Miguel Lezama <miguel@automattic.com>

roccotripaldi committed:
Update _inc/connect-button.js

Co-Authored-By: Miguel Lezama <miguel@automattic.com>

roccotripaldi committed:
Update _inc/connect-button.js

Co-Authored-By: Miguel Lezama <miguel@automattic.com>

roccotripaldi committed:
cleanup. thanks @lezama

delawski commented on _inc/connect-button.js:
```diff
@@ -1,6 +1,55 @@
+		},
+		handleAuthorizationComplete: function() {
+			jetpackConnectButton.isRegistering = false;
+			$( '.jp-connect-full__button-container' ).html(
```
I think that with the recent update to the connection flow (D31768-code) we don't need to replace the `iframe` with a status message - the connection status is displayed and updated directly inside the `iframe`.

tyxla commented on _inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -1089,6 +1099,33 @@ public static function disconnect_site( $request ) {
+	 */
+	public static function register_site( $request ) {
+		if ( ! wp_verify_nonce( $request->get_param( 'registration_nonce' ), 'jetpack-registration-nonce' ) ) {
+			return new WP_Error( 'build_connect_url_failed', esc_html__( 'Unable to build the connect URL.  Please reload the page and try again.', 'jetpack' ), array( 'status' => 400 ) );
```
Do we need to localize this error message? If not, we definitely won't need to escape it either.

tyxla commented on class.jetpack-connection-banner.php:
```diff
@@ -134,13 +134,34 @@ public static function enqueue_connect_button_scripts() {
 			'jetpack-connect-button',
 			Assets::get_file_url_for_environment(
-				'_inc/connect-button.js', // TODO - minify?
+				'_inc/connect-button.js',
```
Any reason not to minify this eventually?

roccotripaldi replied:
that comment was removed in error. we will minify on the next iteration and i'll add an asana card!

delawski commented on _inc/connect-button.js:
```diff
@@ -1,6 +1,55 @@
+		isRegistering: false,
+		handleClick: function() {
+			jetpackConnectButton.isRegistering = true;
+			$( '.jp-connect-button' ).text( jpConnect.buttonTextRegistering );
```
Should we disable the button after clicking and changing the label to 'Loading'? Something like this:

```js
$( '.jp-connect-button' ).attr( 'disabled', true );
```

This way the connection flow here would match analogical flows elsewhere, e.g. in Calypso.

tyxla commented on scss/jetpack-connect.scss:
```diff
@@ -0,0 +1,28 @@
+// ==========================================================================
+
+.jp-jetpack-connect__tos-blurb {
+  font-size: .75rem;
```
Does it make sense to make the styles the same as the `jp-banner__tos-blurb`? Namely:

```
.jp-banner__tos-blurb {
	display: block;
	padding: rem( 6px ) 0;
	line-height: 1.5;
	font-size: rem( 11px );
	color: $gray-dark;
	@include minbreakpoint(phablet) {
		margin-left: rem(18px);
	}

	a {
		color: inherit;
	}
}
```

delawski replied:
I don't think so in this case. Basically all of those styles are already inherited by `.jp-jetpack-connect__tos-blurb` and we don't want to have a left margin on any breakpoint.

The only difference is the `font-size` which is `0.75rem => 12px`. The size of 12 pixels is taken from the Figma mock up.

I think that as long as it looks and works as expected it doesn't make sense add up new CSS rules, but I'm happy to adjust.

tyxla replied:
Happy to keep it as-is then 👍

tyxla commented on scss/jetpack-connect.scss:
```diff
@@ -0,0 +1,28 @@
+
+.jp-jetpack-connect__iframe {
+  height: 240px;
+  width: 380px;
```
Seems like we need varying height here and there. Should we have different classes for that?

delawski replied:
The iframe size is supposed to stay mostly fixed with D31768-code checked out. The only variation is for mobile vs. desktop (reflected in the media query below).

tyxla replied:
Thanks for explaining 👍

tyxla reviewed (changes requested):
This is wonderful work! 💯 

I was never able to finish the process - after authorization I never got redirected to the admin so I could see the connection was finished. It's worth noting that after manually refreshing, I was connected, so I guess we're only failing to redirect for some reason.

Also, left some questions, comments and observations, nothing blocking really. Once we fix the redirect, I think this should be GTG!

delawski commented on _inc/connect-button.js:
```diff
@@ -1,6 +1,55 @@
+			) {
+				if ( event.data === 'close' ) {
+					window.removeEventListener( 'message', this.receiveData );
+					jetpackConnectButton.handleAuthorizationComplete();
```
@tyxla Have you tried checking out D31768-code? It may fix the issues you're reporting here.

tyxla replied:
Nope, I didn't - I tested D31768-code with the original branch Dan worked in. I'm doubtful it would fix it though 🤔

delawski commented on _inc/connect-button.js:
```diff
@@ -1,6 +1,55 @@
+				success: function( data ) {
+					window.addEventListener( 'message', jetpackConnectButton.receiveData );
+					jetpackConnectIframe.attr( 'src', data.authorizeUrl );
+					$( '.jp-connect-full__button-container' ).html( jetpackConnectIframe );
```
@tyxla Have you tried checking out D31768-code? It may fix the issues you're reporting here.

tyxla replied:
Good point, I think it could! The frame was higher when testing with this diff.

tyxla reviewed (dismissed):
Noting that we got redirection to work 👍 The other notes I left are minor and non-blocking. Feel free to ship at will!

delawski committed:
Fix minor issues with connect in place styling and flow.
- Remove 'Finishing up' message after successful authorization in favor of a status indicator inside the `iframe`.
- Disable 'Set up Jetpack' button after click.
- Remove 'Set up Jetpack' button's overrides introduced in `jetpack-connect` (green to blue).
- Fix minor whitespace issues in `jetpack-connection-banner`.

delawski committed:
Combine two conditionals into a single statement.

delawski committed:
Update _inc/lib/class.core-rest-api-endpoints.php

Co-Authored-By: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

delawski commented:
@roccotripaldi @tyxla I've pushed two commits to address some of the feedback left in the code review and to implement the updates I mentioned at our meeting. Please have another look when you get a chance.

roccotripaldi committed:
Use a dynamic API url

roccotripaldi committed:
Improve error messaging.

roccotripaldi committed:
do not use the iframe within banners

tyxla commented:
Great work, folks 🙌
